tests/utils/validate_docstring_examples.py

```diff
@@ -0,0 +1,212 @@
+# License: MIT
+# Copyright © 2022 Frequenz Energy-as-a-Service GmbH
+
+"""Validate docstring code examples."""
+
+from __future__ import annotations
+
+import argparse
+import importlib.util
+import inspect
+import re
+import sys
+import textwrap
+import types
+from pathlib import Path
+
+
+def get_module_docs(module: types.ModuleType) -> list[str]:
+    """Get all docstrings found in the given module.
+
+    Args:
+        module: The module to extract docstrings from.
+
+    Returns:
+        A list of docstrings.
+    """
+    docs = []
+
+    for _, obj in inspect.getmembers(module):
+        if inspect.isfunction(obj) and obj.__module__ == module.__name__:
+            if obj.__doc__ is not None:
+                docs.append(obj.__doc__)
+        elif inspect.isclass(obj) and obj.__module__ == module.__name__:
+            if obj.__doc__ is not None:
+                docs.append(obj.__doc__)
+            for _, method_obj in inspect.getmembers(obj):
+                if (
+                    inspect.isfunction(method_obj)
+                    and method_obj.__module__ == module.__name__
+                ):
+                    if method_obj.__doc__ is not None:
+                        docs.append(method_obj.__doc__)
```

Comment:
From what I learned only a few symbols have the __doc__ property, namely modules, functions, methods and classes.

Well, that's not exactly correct, vars have the property, but it contains type documentation rather than user documentation.

E.g.

```
>>> x = 3
>>> dir(x)
['__abs__', '__add__', '__and__', '__bool__', '__ceil__', '__class__', '__delattr__', '__dir__', '__divmod__', '__doc__', '__eq__', '__float__', '__floor__', '__floordiv__', '__format__', '__ge__', '__getattribute__', '__getnewargs__', '__gt__', '__hash__', '__index__', '__init__', '__init_subclass__', '__int__', '__invert__', '__le__', '__lshift__', '__lt__', '__mod__', '__mul__', '__ne__', '__neg__', '__new__', '__or__', '__pos__', '__pow__', '__radd__', '__rand__', '__rdivmod__', '__reduce__', '__reduce_ex__', '__repr__', '__rfloordiv__', '__rlshift__', '__rmod__', '__rmul__', '__ror__', '__round__', '__rpow__', '__rrshift__', '__rshift__', '__rsub__', '__rtruediv__', '__rxor__', '__setattr__', '__sizeof__', '__str__', '__sub__', '__subclasshook__', '__truediv__', '__trunc__', '__xor__', 'as_integer_ratio', 'bit_count', 'bit_length', 'conjugate', 'denominator', 'from_bytes', 'imag', 'numerator', 'real', 'to_bytes']
>>> x.__doc__
"int([x]) -> integer\nint(x, base=10) -> integer\n\nConvert a number or string to an integer, or return 0 if no arguments\nare given.  If x is a number, return x.__int__().  For floating point\nnumbers, this truncates towards zero.\n\nIf x is not a number or if base is given, then x must be a string,\nbytes, or bytearray instance representing an integer literal in the\ngiven base.  The literal can be preceded by '+' or '-' and be surrounded\nby whitespace.  The base defaults to 10.  Valid bases are 0 and 2-36.\nBase 0 means to interpret the base from the string as an integer literal.\n>>> int('0b100', base=0)\n4"
```

Reply:
Ah, interesting. Well, the docs extracting tool ([mkdocstrings](https://mkdocstrings.github.io/)) is considering strings following a variable a docstring, I guess it does it by parsing the Python file then.

I guess for the first version it is OK to ignore them, but we should support it in the future.

This also makes me think that testing the documentation should probably be part of building the documentation (or an option), so maybe in the future we could plug it to mkdocstring (making it a plugin or something). This will also ensure that the documentation we get to extract examples is the same as the one that will be rendered. The guy doing mkdocstrings seems very nice, he even replied to some issue in our repo when discussing about it.

Reply:
How is it with sybil? I guess it also not considering strings after a variable declaration a docstring for the variable, right?

Reply:
I believe it is supported, as Sybil doesn't use the `__doc__` property but parses the files itself